tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,397 @@
+            else:
+                cmd_line.extend(['-sess_out', self.options.ticket_file])
+        return cmd_line
+
+    def setup_server(self):
+        cmd_line = OpenSSLBase.setup_server(self)
+        cmd_line.extend(['-early_data'])
+        return cmd_line
+
+
+# The `-reconnect` option is broken for TLS1.3 in OpenSSL s_client: https://github.com/openssl/openssl/issues/8517
+# The `-sess_in`/`-sess_out` options can be used instead, but don't have an s2nc equivalent.
+# As we add more providers, we may need both a `-reconnect`-like and a `-sess_in/out`-like S2N server test,
+# but for now we can just use `-sess_in/out` and cover the S2N->S2N case in the S2N client tests.
+CLIENT_PROVIDERS = [ OpenSSL ]
+SERVER_PROVIDERS = [ OpenSSL, S2N ]
+
+
+def get_early_data_bytes(file_path, early_data_size):
+    early_data = data_bytes(early_data_size)
+    with open(file_path, 'wb') as fout:
+        fout.write(early_data)
+    return early_data
+
+
+def get_ticket_from_s2n_server(options, managed_process, provider, certificate):
+    port = next(available_ports)
+
+    """
+    Generally clients start checking for stdin EoF to exit as soon as they finish the handshake.
+    To make sure the client reliably receives the post-handshake NST,
+    do NOT indicate stdin EoF until after some data has been received from the server.
+    """
+    close_marker_bytes = data_bytes(10)
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+    client_options.port = port
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.port = port
+    server_options.key = certificate.key
+    server_options.cert = certificate.cert
+    server_options.data_to_send = close_marker_bytes
+
+    assert not os.path.exists(options.ticket_file)
+
+    s2n_server = managed_process(S2N, server_options, send_marker=S2N.get_send_marker())
+    client = managed_process(provider, client_options, close_marker=str(close_marker_bytes))
+
+    for results in s2n_server.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+
+    for results in client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+
+    assert os.path.exists(options.ticket_file)
+
+
+"""
+Basic S2N server happy case.
+
+We make one full connection to get a session ticket with early data enabled,
+then another resumption connection with early data.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", CLIENT_PROVIDERS, ids=get_parameter_name)
```

Comment:
I think that will add confusion, since we don't do if for any other integration test, even though all tests either use the provider as a client xor server.

Reply:
And it'll definitely break invalid_test_parameters